runtime/src/cost_tracker.rs

```diff
@@ -180,8 +189,17 @@ impl CostTracker {
             return Err(CostTrackerError::WouldExceedAccountMaxLimit);
         }
 
-        if self.account_data_size.saturating_add(account_data_len) > MAX_ACCOUNT_DATA_LEN {
-            return Err(CostTrackerError::WouldExceedAccountDataMaxLimit);
+        // NOTE: Check if the total accounts data size is exceed *before* the block accounts data
+        // size.  This way, transactions are not unnecessarily retried.
+        let account_data_size = self.account_data_size.saturating_add(account_data_len);
+        if let Some(account_data_size_limit) = self.account_data_size_limit {
+            if account_data_size > account_data_size_limit {
+                return Err(CostTrackerError::WouldExceedAccountDataTotalLimit);
+            }
+        }
+
+        if account_data_size > MAX_ACCOUNT_DATA_LEN {
```

Comment:
nit: rename this constant to `MAX_ACCOUNT_DATA_BLOCK_LEN`

Reply:
Done!